pkg/drivers/krunkit/krunkit.go

```diff
@@ -0,0 +1,561 @@
+	return os.OpenFile(logfile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+}
+
+func (d *Driver) setupIP(mac string) error {
+	var err error
+	getIP := func() error {
+		d.IPAddress, err = pkgdrivers.GetIPAddressByMACAddress(mac)
+		if err != nil {
+			return errors.Wrap(err, "failed to get IP address")
+		}
+		return nil
+	}
+	// Implement a retry loop because IP address isn't added to dhcp leases file immediately
+	for i := 0; i < 60; i++ {
+		log.Debugf("Attempt %d", i)
+		err = getIP()
+		if err == nil {
+			break
+		}
+		time.Sleep(2 * time.Second)
+	}
+
+	if err == nil {
+		log.Debugf("IP: %s", d.IPAddress)
+		return nil
+	}
+	if !isBootpdError(err) {
+		return errors.Wrap(err, "IP address never found in dhcp leases file")
+	}
+	if unblockErr := firewall.UnblockBootpd(); unblockErr != nil {
+		klog.Errorf("failed unblocking bootpd from firewall: %v", unblockErr)
+		exit.Error(reason.IfBootpdFirewall, "ip not found", err)
+	}
+	out.Styled(style.Restarting, "Successfully unblocked bootpd process from firewall, retrying")
+	return fmt.Errorf("ip not found: %v", err)
+}
+
+func isBootpdError(err error) bool {
+	return strings.Contains(err.Error(), "could not find an IP address")
+}
+
+func (d *Driver) Stop() error {
+	if err := d.stopKrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Stop()
+}
+
+func (d *Driver) Kill() error {
+	if err := d.killkrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Kill()
+}
+
+func (d *Driver) Remove() error {
+	s, err := d.GetState()
+	if err != nil {
+		return errors.Wrap(err, "get state")
+	}
+	if s == state.Running {
+		if err := d.Kill(); err != nil {
+			return errors.Wrap(err, "kill")
+		}
+	}
+	return nil
+}
+
+func (d *Driver) Restart() error {
+	s, err := d.GetState()
+	if err != nil {
+		return err
+	}
+
+	if s == state.Running {
+		if err := d.Stop(); err != nil {
+			return err
+		}
+	}
+	return d.Start()
+}
+
+func (d *Driver) StartDocker() error {
+	return fmt.Errorf("hosts without a driver cannot start docker")
+}
+
+func (d *Driver) StopDocker() error {
+	return fmt.Errorf("hosts without a driver cannot stop docker")
+}
+
+func (d *Driver) GetDockerConfigDir() string {
+	return ""
+}
+
+func (d *Driver) Upgrade() error {
+	return fmt.Errorf("hosts without a driver cannot be upgraded")
+}
+
+func (d *Driver) getKrunkitState() (state.State, error) {
+	pidfile := d.pidfilePath()
+	pid, err := process.ReadPidfile(pidfile)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return state.Error, err
+		}
+		return state.Stopped, nil
+	}
+	exists, err := process.Exists(pid, driverName)
+	if err != nil {
+		return state.Error, err
+	}
+	if !exists {
+		// No process, stale pidfile.
+		if err := os.Remove(pidfile); err != nil {
+			log.Debugf("failed to remove %q: %s", pidfile, err)
+		}
+		return state.Stopped, nil
+	}
+	return state.Running, nil
+}
+
+func (d *Driver) stopKrunkit() error {
+	if err := d.SetKrunkititState("Stop"); err != nil {
```

Comment:
krunkit ignores this value, but it is documented, so we should keep it same as vfkit.